File: native/cpp/fuzzy.cpp
// Mochi native fuzzy line matcher (C++).
//
// Same contract as TypeScript fuzzyFindUnique in src/tools/fuzzy-match.ts and
// the Rust build in native/rust/fuzzy.rs. Finds the unique region in `text`
// that matches `needle` after whitespace normalization; multiple matches are
// ambiguous and return NONE.
//
// Wire protocol (stdin lines -> stdout one line):
//   first line:  "NEEDLE"
//   needle lines...
//   a line with exactly "---TEXT---" separates needle from text
//   text lines...
// Output: OK <start> <end>  |  NONE  |  ERR <message>
//
// Build: g++ -O2 -std=c++17 -o bin/fuzzy_cpp native/cpp/fuzzy.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

static std::string normalize(const std::string& line) {
    std::istringstream in(line);
    std::string word;
    std::vector<std::string> words;
    while (in >> word) words.push_back(word);
    std::string out;
    for (size_t i = 0; i < words.size(); ++i) {
        if (i) out += ' ';
        out += words[i];
    }
    return out;
}

static bool find_unique(const std::vector<std::string>& needle,
                        const std::vector<std::string>& text,
                        long long* out_start, long long* out_end) {
    if (needle.empty() || needle.size() > text.size()) return false;
    std::vector<std::string> n = needle, t;
    for (const auto& line : text) t.push_back(normalize(line));
    for (auto& line : n) line = normalize(line);

    long long count = 0, start = 0, end = 0;
    for (size_t i = 0; i + n.size() <= t.size(); ++i) {
        bool ok = true;
        for (size_t j = 0; j < n.size(); ++j) {
            if (t[i + j] != n[j]) { ok = false; break; }
        }
        if (!ok) continue;
        long long s = 0;
        for (size_t k = 0; k < i; ++k) s += (long long)text[k].size() + 1;
        long long e = s;
        for (size_t k = i; k < i + n.size(); ++k) e += (long long)text[k].size() + 1;
        if (e > s) --e;
        ++count;
        if (count == 1) { start = s; end = e; }
    }
    if (count == 1) { *out_start = start; *out_end = end; return true; }
    return false;
}

int main() {
    std::string line;
    std::vector<std::string> lines;
    while (std::getline(std::cin, line)) lines.push_back(line);

    std::vector<std::string> needle, text;
    bool in_text = false;
    for (const auto& l : lines) {
        if (l == "---TEXT---" || l == "---TEXT---\r") { in_text = true; continue; }
        if (in_text) text.push_back(l);
        else needle.push_back(l);
    }
    // Drop the leading "NEEDLE" marker line if present.
    if (!needle.empty() && needle[0] == "NEEDLE") needle.erase(needle.begin());
    // Drop empty leading/trailing needle lines, matching the TS impl.
    while (!needle.empty() && needle.front().empty()) needle.erase(needle.begin());
    while (!needle.empty() && needle.back().empty()) needle.pop_back();

    if (needle.empty()) { std::cout << "NONE\n"; return 0; }
    long long s = 0, e = 0;
    if (find_unique(needle, text, &s, &e)) {
        std::cout << "OK " << s << " " << e << "\n";
    } else {
        std::cout << "NONE\n";
    }
    return 0;
}

File: native/rust/fuzzy.rs
// Mochi native fuzzy line matcher (Rust).
//
// Same contract as TypeScript fuzzyFindUnique in src/tools/fuzzy-match.ts:
// finds the unique region in `text` that matches `needle` after whitespace
// normalization. Multiple matches => ambiguous => NONE, so the agent never
// edits the wrong occurrence.
//
// Wire protocol (stdin lines -> stdout one line):
//   first line:  "NEEDLE"
//   needle lines...
//   a line with exactly "---TEXT---" separates needle from text
//   text lines...
// Output: OK <start> <end>  |  NONE  |  ERR <message>
//
// Build: rustc -O -o bin/fuzzy_rust native/rust/fuzzy.rs
use std::io::{self, Read};

/// Collapse runs of whitespace to a single space (and trim), matching the TS
/// normalizeLine(): `line.trim().replace(/\s+/g, ' ')`.
fn normalize(line: &str) -> String {
    line.split_whitespace().collect::<Vec<_>>().join(" ")
}

fn find_unique_range(needle: &[String], text: &[String]) -> Option<(usize, usize)> {
    if needle.is_empty() || needle.len() > text.len() {
        return None;
    }
    let norm_needle: Vec<String> = needle.iter().map(|l| normalize(l)).collect();
    let norm_text: Vec<String> = text.iter().map(|l| normalize(l)).collect();

    let mut count = 0usize;
    let mut found = (0usize, 0usize);
    for i in 0..=(norm_text.len() - norm_needle.len()) {
        let mut ok = true;
        for j in 0..norm_needle.len() {
            if norm_text[i + j] != norm_needle[j] {
                ok = false;
                break;
            }
        }
        if !ok {
            continue;
        }
        let mut start = 0usize;
        for k in 0..i {
            start += text[k].len() + 1;
        }
        let mut end = start;
        for k in i..(i + norm_needle.len()) {
            end += text[k].len() + 1;
        }
        if end > start {
            end -= 1;
        }
        count += 1;
        if count == 1 {
            found = (start, end);
        }
    }
    if count == 1 {
        Some(found)
    } else {
        None
    }
}

fn main() {
    let mut input = String::new();
    if io::stdin().read_to_string(&mut input).is_err() {
        println!("ERR read-failed");
        return;
    }
    let mut lines: Vec<String> = input.split('\n').map(|s| s.to_string()).collect();
    // Protocol: first line is the marker "NEEDLE" (ignored); then needle,
    // then separator, then text.
    let mut needle: Vec<String> = Vec::new();
    let mut text: Vec<String> = Vec::new();
    let mut in_text = false;
    for line in lines.drain(..) {
        if line.trim() == "---TEXT---" {
            in_text = true;
            continue;
        }
        if in_text {
            text.push(line);
        } else {
            needle.push(line);
        }
    }
    // Drop the leading "NEEDLE" marker line if present.
    if needle.first().map(|l| l.trim()) == Some("NEEDLE") {
        needle.remove(0);
    }
    // Drop empty leading/trailing needle lines, matching the TS impl.
    while needle.first().map(|l| l.trim().is_empty()) == Some(true) {
        needle.remove(0);
    }
    while needle.last().map(|l| l.trim().is_empty()) == Some(true) {
        needle.pop();
    }
    if needle.is_empty() {
        println!("NONE");
        return;
    }
    match find_unique_range(&needle, &text) {
        Some((s, e)) => println!("OK {} {}", s, e),
        None => println!("NONE"),
    }
}